pkg/operator/sync.go

```diff
@@ -198,6 +198,23 @@ func (optr *Operator) syncMutatingWebhook() error {
 }
 
 func (optr *Operator) syncBaremetalControllers(config *OperatorConfig, configName string) error {
+	// Stand down if cluster-bare-metal operator has claimed the metal3 deployment
+	// and if the baremetal clusteroperator exists
+	maoOwned, err := checkMetal3DeploymentMAOOwned(optr.kubeClient.AppsV1(), config)
```

Comment:
This check seems redundant.  If the annotation is being applied by the BMO, there's no need to check for this.  If the BMO exists, then it owns the deployment and should implement GetOrCreate for the deployment.

Reply:
This code is trying to ensure a smooth hand-off between the MAO and CBO during upgrade and downgrade. The details are described in the enhancement document at https://github.com/openshift/enhancements/blob/master/enhancements/baremetal/an-slo-for-baremetal.md#upgrade--downgrade--version-skew-strategy

Reply:
We definitely need this change when we downgrade from a 4.7 MAO which does not manage metal3 to 4.6 MAO that does. We are achieving that by checking for the presence of the cboOwnedAnnotation and the presence of CBO's ClusterOperator.

Reply:
In 4.7, it will be removed along with the rest of baremetal code in MAO.

Reply:
I disagree this is needed.

Scenario in 4.6:  CBO not deployed, MAO owned, done.
Scenario in 4.7:  CBO gets deployed at some point, MAO doesn't own, MAO quits syncing.
Scenario in 4.7 -> 4.6 downgrade: CBO is initially deployed, MAO does nothing. CBO goes away, MAO starts syncing again.

There's not a scenario that needs an annotation that I can see.

Reply:
The 4.6 and 4.7 configurations for metal3 will be different. Therefore if we ever have a time when 2 components are both trying to update the Deployment, they're going to thrash making changes back and forth. There are at least 2 situations where that could come up.

First, during a production upgrade from 4.6 to 4.7 we cannot ensure the order in which the CVO launches the new version of the MAO and the CBO. If the CBO launches before the new MAO, then the definition of the metal3 Deployment would thrash until the MAO is updated. Restarting that pod is extremely disruptive. It has a DHCP server and a relational database and a bunch of other unusual stuff in it. We can't prevent it from restarting entirely, but we want to minimize the restarts, especially in production clusters.

Second, the CBO is not yet fully functional. We want to build it incrementally. As part of doing that, we need a way to run integration test jobs in which the CBO manages metal3 and the MAO does not, so we can ensure we can deploy new clusters from scratch just with the CBO. That test job will tell us when we have the CBO working well enough to be able to remove the metal3 logic from the MAO entirely. Without this change, it would not be possible to have that test job because the 2 components will compete to control the Deployment and we wouldn’t have a way to ensure that the CBO is the actor causing the right outcome (if the Deployment was even stable long enough to fully launch a Pod). We also couldn’t have a non-flakey 4.6->4.7 upgrade test job for bare metal, because we would run into the upgrade sequencing issue described in the previous paragraph.

We did consider some other approaches.

For example, we could wait until near the end of 4.7 and just remove the metal3 management code from the MAO after the CBO is thought to work. But, that leaves us with several sprints/months while the CBO is under development during which it would not be possible to have a stable metal cluster for development because of the thrashing. We also couldn't have those CI jobs. So, we would be waiting until very late in the cycle to do any integration testing for a significant change to the system.

We could also remove that metal3 management code from the MAO as soon as 4.7 opens. That would be even more disruptive, because it would leave us without the ability to stand up metal clusters *at all*, until the CBO is fully functional, because the MAO would no longer launch the metal3 components needed to complete the IPI installation. No one working with bare metal clusters would be able to get any work done until the new component was fully functional.

This is a relatively small change. Nothing in 4.6 is going to set the annotation so it should have no effect (outside of development/CI) until an upgrade occurs. We anticipate being able to remove this logic from the MAO in 4.7 at the same time that we remove the rest of the metal3 management logic, after the CBO is fully functional. But, we think it is needed for now, and the sooner we get it into 4.6 the more likely it will be that we can avoid forcing users to upgrade through a specific 4.6 version before going to 4.7.

If you have other suggestions to address the production upgrade sequencing issue and the incremental development requirement, please let us know.

Reply:
> Second, the CBO is not yet fully functional. We want to build it incrementally. As part of doing that, we need a way to run integration test jobs in which the CBO manages metal3 and the MAO does not, so we can ensure we can deploy new clusters from scratch just with the CBO. 

Okay, this makes some sense then.

My only other concern is there are tests for checking if both annotations are on the deployment, but nothing in the MAO making that check.  I don't think there's a scenario where this would actually happen.  Perhaps instead of a new CBO annotation, we should utilize the k/v of a singular annotation such as
"baremetal.openshift.io/managed-by: mao" / "baremetal.openshift.io/managed-by: cbo".  This will prevent PATCH updates from either source resulting in both annotations being present.

Reply:
checkMetal3DeploymentMAOOwned() givens preference to the cboOwnedAnnotation when present and if not, assumes it is owned by MAO. So, with that we are in effect taking care of the case where if both cboOwnedAnnotation and maoOwnedAnnotation are present on a deployment, the cboOwnedAnnotation takes precedence.

Reply:
We could use the k/v of a single annotation just as you pointed out. But, that would essentially ignore the pre-existing annotation added to any deployment owned by MAO. We followed what was done as part of a previous MAO feature https://github.com/openshift/machine-api-operator/pull/424 and used a similar annotation for CBO.

Reply:
>First, during a production upgrade from 4.6 to 4.7 we cannot ensure the order in which the CVO launches the new version of the MAO and the CBO. 

This could be ensured via prefix https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/upgrades.md#upgrades-and-order

>Second, the CBO is not yet fully functional. We want to build it incrementally...

This make sense to me. I think checking CBO existence  should be enough.

Reply:
> > First, during a production upgrade from 4.6 to 4.7 we cannot ensure the order in which the CVO launches the new version of the MAO and the CBO.
> 
> This could be covered and ensured via prefix https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/upgrades.md#upgrades-and-order

Yeah, but it's pretty ugly to put CBO in a later runlevel just for this temporary upgrade handover. Even that link suggests using another mechanism - "Over time, operators should be better at detecting their prerequisites without overcomplicating or risking the predictability of upgrades"